Sources/GRPCInProcessTransport/InProcessClientTransport.swift

```diff
@@ -281,56 +281,64 @@ public final class InProcessClientTransport: ClientTransport {
           connectedState.openStreams[streamID] = (clientStream, serverStream)
           connectedState.nextStreamID += 1
           state = .connected(connectedState)
+          return .success(streamID)
         } catch let acceptStreamError as RPCError {
-          serverStream.outbound.finish(throwing: acceptStreamError)
-          clientStream.outbound.finish(throwing: acceptStreamError)
-          throw acceptStreamError
+          return .failure(acceptStreamError)
         } catch {
-          serverStream.outbound.finish(throwing: error)
-          clientStream.outbound.finish(throwing: error)
-          throw RPCError(code: .unknown, message: "Unknown error: \(error).")
+          return .failure(RPCError(code: .unknown, message: "Unknown error: \(error)."))
         }
-        return streamID
 
       case .closed:
-        let error = RPCError(
-          code: .failedPrecondition,
-          message: "The client transport is closed."
-        )
-        serverStream.outbound.finish(throwing: error)
-        clientStream.outbound.finish(throwing: error)
-        throw error
+        let error = RPCError(code: .failedPrecondition, message: "The client transport is closed.")
+        return .failure(error)
       }
     }
 
-    defer {
-      clientStream.outbound.finish()
-
-      let maybeEndContinuation = self.state.withLock { state in
-        switch state {
-        case .unconnected:
-          // The state cannot be unconnected at this point, because if we made
-          // it this far, it's because the transport was connected.
-          // Once connected, it's impossible to transition back to unconnected,
-          // so this is an invalid state.
-          fatalError("Invalid state")
-        case .connected(var connectedState):
-          connectedState.openStreams.removeValue(forKey: streamID)
-          state = .connected(connectedState)
-        case .closed(var closedState):
-          closedState.openStreams.removeValue(forKey: streamID)
-          state = .closed(closedState)
-          if closedState.openStreams.isEmpty {
-            // This was the last open stream: signal the closure of the client.
-            return closedState.signalEndContinuation
-          }
-        }
-        return nil
+    switch acceptStream {
+    case .success(let streamID):
+      let streamHandlingResult: Result<T, any Error>
+      do {
+        let result = try await closure(clientStream)
+        streamHandlingResult = .success(result)
+      } catch {
+        streamHandlingResult = .failure(error)
```

Comment:
We didn't do that previously because the server is responsible for its stream. We only finish it in the failure case because the server never accepted the stream so the client is still responsible for it.